uber/models/attendee.py

```diff
@@ -826,33 +833,74 @@ def hour_map(self):
+            .difference(member_dept_ids)
+
+        member_filter = Job.department_id.in_(member_dept_ids) \
+            if member_dept_ids else None
+
+        max_depts = c.MAX_DEPTS_WHERE_WORKING
+        no_max = max_depts < 1
+
+        requested_filter = None
+        if requested_dept_ids and (len(member_dept_ids) < max_depts or no_max):
+            depts_where_working = set(j.department_id for j in self.jobs)
+            if len(depts_where_working) >= max_depts and not no_max:
+                requested_dept_ids = depts_where_working \
+                    .difference(member_dept_ids)
+
+            requested_any_dept = None in requested_dept_ids
+            if requested_any_dept:
+                requested_filter = Job.visibility >= self.job_visibility
+            elif requested_dept_ids:
+                requested_filter = and_(
+                    Job.visibility >= self.job_visibility,
+                    Job.department_id.in_(requested_dept_ids))
+
+        if member_filter is not None and requested_filter is not None:
+            job_filters += [or_(member_filter, requested_filter)]
+        elif member_filter is not None:
+            job_filters += [member_filter]
+        elif requested_filter is not None:
+            job_filters += [requested_filter]
+        return job_filters
+
     @cached_property
     def available_jobs(self):
-        if not self.dept_memberships and not c.AT_THE_CON:
-            return []
+        assert self.session, (
+            '{}.available_jobs property may only be accessed for '
+            'objects attached to a session'.format(self.__class__.__name__))
 
-        def _get_available_jobs(session, attendee_id):
-            from uber.models.department import DeptMembership, Job, Shift
-            job_filters = [] if c.AT_THE_CON else [
-                Job.department_id == DeptMembership.department_id,
-                DeptMembership.attendee_id == self.id]
-            return session.query(Job) \
-                .outerjoin(Job.shifts) \
-                .filter(*job_filters) \
-                .group_by(Job.id) \
-                .having(func.count(Shift.id) < Job.slots) \
-                .options(
-                    subqueryload(Job.shifts),
-                    subqueryload(Job.required_roles)) \
-                .order_by(Job.start_time, Job.department_id).all()
-
-        if self.session:
-            jobs = _get_available_jobs(self.session, self.id)
-        else:
-            from uber.models import Session
-            with Session() as session:
-                jobs = _get_available_jobs(session, self.id)
+        if not self.staffing or (
+                not c.AT_THE_CON and
+                not self.dept_memberships and
+                not self.dept_membership_requests):
+            return []
 
+        from uber.models.department import Job
+        job_filters = self.available_job_filters
+        jobs = self.session.query(Job).filter(*job_filters) \
```

Comment:
Since you only use it once, I would get rid of `job_filters` and just say `.filter(*self.available_job_filters)`. Either way is fine, of course!

Reply:
Haha, I think that was a holdover from some earlier code that repeatedly mentioned `job_filters`. But yes, it would be cleaner the other way.